pep-0654.rst

```diff
@@ -335,37 +335,111 @@ typically either query to check if it has leaf exceptions for which some
 interested in the number of ``Ts`` in ``eg``.
 
-If it does turn out to be necessary for an application to iterate over the
-individual exceptions of an ``ExceptionGroup`` ``eg``, this can be done by
-calling ``traverse(eg)``, where ``traverse`` is defined as follows:
+However, there are situations where it is necessary to inspect the
+individual leaf exceptions. For example, suppose that we have an
+``ExceptionGroup`` ``eg`` and we want to log the ``OSErrors`` that have a
+specific error code and reraise everything else. We can do this by passing
+a function with side effects to ``subgroup``, as follows:
 
 .. code-block::
 
-    def traverse(exc, tbs=None):
+    def log_and_ignore_ENOENT(err):
+        if isinstance(err, OSError) and err.errno == ENOENT:
+            log(err)
+            return False
+        else:
+            return True
+
+    try:
+        . . .
+    except ExceptionGroup as eg:
+        eg = eg.subgroup(log_and_ignore_ENOENT)
+        if eg is not None:
+            raise eg
+
+
+In the previous example, when ``log_and_ignore_ENOENT`` is invoked on a leaf
+exception, only part of this exception's traceback is accessible - the part
+referenced from its ``__traceback__`` field. If we need the full traceback,
+we need to look at the concatenation of the tracebacks of the exceptions on
+the path from the root to this leaf. we can get that with direct iteration,
+recursively, as follows:
+
+.. code-block::
+
+    def leaf_generator(exc, tbs=None):
         if tbs is None:
             tbs = []
 
         tbs.append(exc.__traceback__)
         if isinstance(exc, ExceptionGroup):
             for e in exc.errors:
-                traverse(e, tbs)
+                yield from leaf_generator(e, tbs)
         else:
             # exc is a leaf exception and its traceback
-            # is the concatenation of the traceback in tbs
-            process_leaf(exc, tbs)
+            # is the concatenation of the traceback
+            # segments in tbs
+            yield exc, tbs
         tbs.pop()
 
 
+We can then process the full tracebacks of the leaf exceptions:
+
+.. code-block::
+
+    >>> import traceback
+    >>>
+    >>> def g(v):
+    ...     try:
+    ...         raise ValueError(v)
+    ...     except Exception as e:
+    ...         return e
+    ...
+    >>> def f():
+    ...     raise ExceptionGroup("eg", [g(1), g(2)])
+    ...
+    >>> try:
+    ...     f()
+    ... except BaseException as e:
+    ...     eg = e
+    ...
+    >>> for (i, (exc, tbs)) in enumerate(leaf_generator(eg)):
+    ...     print(f"\n>>> Exception #{i+1}:")
+    ...     traceback.print_exception(exc)
+    ...     print(f"The complete traceback for Exception #{i+1}:")
+    ...     for tb in tbs:
+    ...         traceback.print_tb(tb)
+    ...
+
+    >>> Exception #1:
+    Traceback (most recent call last):
+      File "<stdin>", line 3, in g
+    ValueError: 1
+    The complete traceback for Exception #1
+    File "<stdin>", line 2, in <module>
```

Comment:
```suggestion
      File "<stdin>", line 2, in <module>
```

Reply:
let me check this

Reply:
Yeah the code gives it indented. I messed it up in the copy-paste.